pennylane/capture/meta_type.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Defines a metaclass for automatic integration of any ``Operator`` with jaxpr program capture.
+
+See ``explanations.md`` for technical explanations of how this works.
+"""
+from typing import Iterable
+from functools import lru_cache
+
+import pennylane as qml
+
+from .switches import plxpr_enabled
+
+has_jax = True
+try:
+    import jax
+except ImportError:
+    has_jax = False
+
+
+@lru_cache  # constrcut the first time lazily
+def _get_abstract_operator():
+    """Create an AbstractOperator once in a way protected from lack of a jax install."""
+    if not has_jax:
+        raise ImportError("Jax is required for plxpr.")
+
+    # TODO: investigate
+    # pennylane/capture/meta_type.py:37:4: W0223: Method 'at_least_vspace' is abstract in class 'AbstractValue' but is not overridden in child class 'AbstractOperator' (abstract-method)
+    # pennylane/capture/meta_type.py:37:4: W0223: Method 'join' is abstract in class 'AbstractValue' but is not overridden in child class 'AbstractOperator' (abstract-method)
+    # pennylane/capture/meta_type.py:37:4: W0223: Method 'update' is abstract in class 'AbstractValue' but is not overridden in child class 'AbstractOperator' (abstract-method)
+
+    class AbstractOperator(jax.core.AbstractValue):
+        """An operator captured into plxpr."""
+
+        def __eq__(self, other):
+            return isinstance(other, AbstractOperator)
+
+        def __hash__(self):
+            return hash("AbstractOperator")
+
+        def _matmul(self, *args):
+            return qml.prod(*args)
+
+        def _mul(self, a, b):
+            return qml.s_prod(b, a)
+
+        def _rmul(self, a, b):
+            return qml.s_prod(b, a)
+
+        def _add(self, a, b):
+            return qml.sum(a, b)
+
+    jax.core.raise_to_shaped_mappings[AbstractOperator] = lambda aval, _: aval
+
+    return AbstractOperator
+
+
+class PLXPRMeta(type):
+    """A metatype that:
+
+    * automatically registers a jax primitive to ``cls._primitive``
+    * Dispatches class creation to ``cls._primitive.bind`` instead of normal class
+    creation when the primitive is defined and plxpr capture is enabled.
+
+    See ``pennylane/capture/explanations.md`` for more detailed information on how this technically
+    works.
+
+    """
+
+    def __init__(cls, *_, **__):
+
+        # Called when constructing a new type that has this metaclass.
```

Comment:
Not entirely sure, but is it "defining" instead of "constructing"?
And maybe "a new class" instead of "a new type"

Reply:
I would call writing something like:
```
class MyOp(Operator):
    pass
```

to be "defining".  This is more the infrastructure around that definition. This takes that definition, and "constructs" a full-fledged type around it.

As for "class" versus "type", we could probably go both way. Seems like the terms are pretty interchangeable with python.

Reply:
>This takes that definition, and "constructs" a full-fledged type around it.

I see, makes sense :)

Reply:
changing to class.